OSX: Handle applicationShouldTerminate

grokys opened the pull request:
## What does the pull request do?

Handles [applicationShouldTerminate](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc) on OSX to correctly shutdown the application when the application is right-clicked and "Quit" selected.

## What is the current behavior?

#6149

## What is the updated/expected behavior with this PR?

Choosing Quit from the icon menu has the same effect as clicking the application Quit menu item.

## How was the solution implemented (if it's not obvious)?

Added some new plumbing as described in #2736 (with slightly updated naming) in order to make this behavior customizable:

- When the Quit menu item is clicked, call `NSApplication terminate`
  - (When the Quit menu item from the application icon is called, the OS will also call `terminate`)
- Add a handler for `applicationShouldTerminate`
- Which raises the managed event `IPlatformLifetimeEventsImpl.ShutdownRequested`
- This event is subscribed from `ClassicDesktopStyleApplicationLifetime`
- Which tries to close all non-owned windows
  - Owned windows will be closed by their owners
- And returns false to cancel the OS-level close if not all windows could be closed

## Fixed issues

Fixes #2736 
Fixes #6149

grokys committed:
Handle applicationShouldTerminate on OSX.

Added some new plumbing to facilitate this:

- A managed event is raised `IPlatformLifetimeEventsImpl.ShutdownRequested`
- This event is subscribed from `ClassicDesktopStyleApplicationLifetime`
- Which tries to close all windows
- And returns false to cancel the OS-level close if not all windows could be closed

grokys committed:
Call terminate from Quit menu.

Unifies the closing logic between clicking on the icon menu Quit and the Quit menu.

grokys committed:
Remove dead code.

Takoooooo commented on src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:
```diff
@@ -123,6 +129,14 @@ public void Dispose()
+        
+        private void ShutdownRequested(object sender, CancelEventArgs e)
+        {
+            foreach (var w in Windows.ToArray())
```
Just curious. What is the purpose of calling ToArray() here?

danwalmsley replied:
Inside the event, the list of windows could be modified. Or even a new window opened, so you need to create a snap shot for it to work correctly.

Takoooooo replied:
Look at the Windows property. You already receive a snapshot from this property when you start iterating.

Gillibald replied:
Y looks like the ToArray isn't needed

grokys replied:
Good catch, I'd not noticed that - fixed.

maxkatz6 commented on native/Avalonia.Native/src/OSX/window.mm:
```diff
@@ -1887,18 +1887,7 @@ -(double) getExtendedTitleBarHeight
-            [window performClose:nil];
-        }
-    }
+    [[NSApplication sharedApplication] terminate:self];
```
As I understand it's possible to cancel termination with applicationShouldTerminate only when NSSupportsSuddenTermination is disabled in plist file.

So, previously it was possible to cancel closeAll (since it was just closing windows), but after this PR it won't be possible unless user disabled NSSupportsSuddenTermination.

Not critical, and I would prefer this behavior for consistency. But it should be kept in mind.

grokys replied:
Hmm, OK, that's not mentioned in the docs for [applicationShouldTerminate ](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc) or [terminate](https://developer.apple.com/documentation/appkit/nsapplication/1428417-terminate?language=objc) and the docs for [NSSupportsSuddenTermination]([https://developer.apple.com/documentation/bundleresources/information_property_list/nssupportssuddentermination?language=objc) specify "whether the _system_ may terminate".

Are your sure that this applies when calling `terminate` manually and not just the OS during shutdown?

maxkatz6 replied:
@grokys I am not sure actually how it works with programmatic terminating. Should be checked.

grokys replied:
Given this a quick test and seems to work as expected. Hopefully I didn't mess it up as I'm still learning OSX stuff, but seems to be good.

grokys committed:
Remove unneeded list copy.

Gillibald committed:
Merge branch 'master' into feature/2736-applicationShouldTerminate

ermolinskya1qa commented on src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:
```diff
@@ -123,6 +129,14 @@ public void Dispose()
+        private void ShutdownRequested(object sender, CancelEventArgs e)
+        {
+            foreach (var w in Windows)
+                w.Close();
```
hey guys! it's kinda breaking the logic that was added in this PR a while ago - https://github.com/AvaloniaUI/Avalonia/pull/5616 
With that logic we are able to cancel closing event from parent level. But current PR do it in a different way - just iterates throught every window and goes straight to the closing and we have no chance to cancel it from parent level

grokys replied:
Yeah it seems that #5616 added the logic to `HandleClosing` which means it only gets run when the user clicks the close button, not when closed programmatically. Was this intended @danwalmsley ?

ermolinskya1qa replied:
at that point HandleClosing triggered for CMD + Q too, not only close button. But this new PR changes the behavior of CMD + Q and it's now the same as quit from dock

grokys replied:
Yeah the real issue is that the logic isn't invoked when calling `Window.Close`

Gillibald replied:
We should still only call close on top level windows right? Because all child windows will be closed by their parent.

grokys replied:
https://github.com/AvaloniaUI/Avalonia/pull/5616 added logic for parent windows to call close on child windows. This logic though was added to `HandleClosing` which reacts to a close notification coming from the OS, which means that calling Close from code bypasses this logic. Surely calling Close from code should query the child windows in the same way as the OS signaling close?

grokys replied:
Actually looks like I was wrong, `Close` should call `Closing` on child windows... Need to investigate more.

grokys replied:
I don't think it breaks the logic from #5616 to be honest, but now a shutdown will try to close _all_ windows in the application, even child windows. I think what you're asking for is that a shutdown only tries to close non-child windows, and that the parent window will be responsible for closing them if it doesn't cancel shutdown.

grokys replied:
Yep, this seems like correct behavior: cross checked with WPF and owned windows are closed by their owners on shutdown. Implemented.

grokys committed:
Only try to close non-owned windows on shutdown.

Owned windows will be closed by their owners.

grokys committed:
Don't try to shutdown if e.Cancel already set.

This allows users to add their own `ShutdownRequested` handlers which can override the lifetime's handler.

grokys committed:
Added ShutdownRequested to IClassicDesktopStyleApplicationLifetime.

grokys committed:
Update ApiCompatBaseline.txt

grokys committed:
Merge branch 'master' into feature/2736-applicationShouldTerminate

grokys committed:
Undo changes to AppCompatBaseline.txt

danwalmsley reviewed (approved):
LGTM

Commit:
Merge branch 'master' into feature/2736-applicationShouldTerminate
